sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/MemoryBufferedSubscriberImpl.java

```diff
@@ -0,0 +1,175 @@
+import java.util.function.Consumer;
+import java.util.function.Function;
+import org.apache.beam.sdk.io.gcp.pubsublite.internal.MemoryLimiter.Block;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MemoryBufferedSubscriberImpl extends ProxyService implements MemoryBufferedSubscriber {
+  private static final Logger LOG = LoggerFactory.getLogger(MemoryBufferedSubscriberImpl.class);
+
+  private final Partition partition;
+  private final MemoryLimiter limiter;
+  private final Subscriber subscriber;
+  private final long maxMemory;
+  private long targetMemory;
+  private Offset fetchOffset;
+  private Block memBlock;
+
+  private long bytesOutstandingToServer = 0;
+  private long bytesOutstanding = 0;
+  private final Queue<SequencedMessage> messages = new ArrayDeque<>();
+  private SettableApiFuture<Void> newData = SettableApiFuture.create();
+  private boolean shutdown = false;
+
+  // onReceive will not be called inline as subscriber is not started.
+  // addServices is intended to be called from the constructor.
+  @SuppressWarnings({"methodref.receiver.bound.invalid", "method.invocation.invalid"})
+  public MemoryBufferedSubscriberImpl(
+      Partition partition,
+      Offset startOffset,
+      MemoryLimiter limiter,
+      Function<Consumer<List<SequencedMessage>>, Subscriber> subscriberFactory,
+      long maxMemory) {
+    this.partition = partition;
+    this.fetchOffset = startOffset;
+    this.limiter = limiter;
+    this.maxMemory = maxMemory;
+    this.targetMemory = maxMemory;
+    this.subscriber = subscriberFactory.apply(this::onReceive);
+    addServices(this.subscriber);
+    memBlock = limiter.claim(targetMemory);
+  }
+
+  @Override
+  protected synchronized void start() throws CheckedApiException {
+    bytesOutstandingToServer += memBlock.claimed();
+    bytesOutstanding += memBlock.claimed();
+    subscriber.allowFlow(
+        FlowControlRequest.newBuilder()
+            .setAllowedBytes(memBlock.claimed())
+            .setAllowedMessages(Long.MAX_VALUE)
+            .build());
+  }
+
+  @Override
+  protected synchronized void stop() {
+    if (shutdown) {
+      return;
+    }
+    shutdown = true;
+    newData.set(null);
+    memBlock.close();
+  }
+
+  @Override
+  protected synchronized void handlePermanentError(CheckedApiException e) {
+    stop();
+  }
+
+  private synchronized void onReceive(List<SequencedMessage> batch) {
+    if (shutdown) {
+      return;
+    }
+    for (SequencedMessage message : batch) {
+      bytesOutstandingToServer -= message.getSizeBytes();
+    }
+    messages.addAll(batch);
+    newData.set(null);
+    newData = SettableApiFuture.create();
+  }
+
+  @Override
+  public synchronized Offset fetchOffset() {
+    return fetchOffset;
+  }
+
+  @Override
+  public synchronized void rebuffer() throws ApiException {
+    if (shutdown) {
+      return;
+    }
+    if (bytesOutstandingToServer < (targetMemory / 3)) {
+      // Server is delivering lots of data
```

Comment:
How about expanding comments here.  
// Server is delivering lots of data, increase the target so that it is not throttled.
and below
// Server is delivering little data, decrease the target so that memory can be used for other users of the limiter.

And if my understanding is wrong even more reason to have comments :)

Reply:
Done.